File: Frontend/angular-app/src/app/modules/bursary/repositories/financial-request/receipt-cheque/commands/create-cheque-sheet-command.ts
import {IRequest, IRequestHandler} from "../../../../../../core/services/mediator/interfaces";
import {HttpRequest} from "../../../../../../core/services/http/http-request";
import {NotificationService} from "../../../../../../shared/services/notification/notification.service";
import {Inject} from "@angular/core";
import {ServiceResult} from "../../../../../../core/models/service-result";
import {HttpService} from "../../../../../../core/services/http/http.service";
import {ApplicationError} from "../../../../../../core/exceptions/application-error";
import {MediatorHandler} from "../../../../../../core/services/mediator/decorator";
import {ValidationRule} from "../../../../../../core/validation/validation-rule";
import {Cheque} from "../../../../entities/cheque";
import { ChequeSheet } from "../../../../entities/cheque-sheet";

export class CreateChequeSheetCommand extends IRequest<CreateChequeSheetCommand, ChequeSheet> {

  public id	: number | undefined =undefined
  public payChequeId	: number | undefined =undefined
  public sheetSeqNumber	 : number | undefined =undefined
  public sheetUniqueNumber	 : string | undefined =undefined
  public title	 : string | undefined =undefined

  public sheetSeriNumber	 : number | undefined =undefined
  public totalCost	 : number | undefined =undefined
  public accountReferenceId	: number | undefined =undefined

  public bankId :number | undefined = undefined
  public bankBranchId	: number | undefined =undefined
  public issueDate	 : number | undefined =undefined
  public receiptDate	 : number | undefined =undefined
  public issuerEmployeeId	: number | undefined =undefined
  public description	 : string | undefined =undefined
  public accountNumber	 : string | undefined =undefined
  public branchName	 : string | undefined =undefined
  public chequeSheetId	 : number | undefined =undefined
  public ownerChequeReferenceId : number | undefined = undefined
  public ownerChequeReferenceGroupId	: number | undefined =undefined

  public receiveChequeReferenceId : number | undefined = undefined
  public receiveChequeReferenceGroupId	: number | undefined =undefined

  public recipientChequeReferenceId : number | undefined = undefined
  public chequeTypeBaseId	: number | undefined = undefined
  public isActive	: boolean | undefined = false
  public approveReceivedChequeSheet	: boolean | undefined = undefined


  constructor() {
    super();
  }

  mapFrom(entity: ChequeSheet): CreateChequeSheetCommand {

    this.mapBasics(entity,this);
    return this;

  }

  mapTo(): ChequeSheet {
    throw new ApplicationError(CreateChequeSheetCommand.name, this.mapTo.name, 'Not Implemented Yet')
  }

  get url(): string {
    return "/bursary/chequeSheet/add";
  }

  get validationRules(): ValidationRule[] {
    return [];
  }
}

@MediatorHandler(CreateChequeSheetCommandHandler.name)
export class CreateChequeSheetCommandHandler implements IRequestHandler<CreateChequeSheetCommand, ChequeSheet> {
  constructor(
    @Inject(HttpService) private _httpService: HttpService,
    @Inject(NotificationService) private _notificationService: NotificationService,
  ) {
  }

  async Handle(request: CreateChequeSheetCommand): Promise<ChequeSheet> {
    let httpRequest: HttpRequest<CreateChequeSheetCommand> = new HttpRequest<CreateChequeSheetCommand>(request.url, request);

    return await this._httpService.Post<CreateChequeSheetCommand, ServiceResult<ChequeSheet>>(httpRequest).toPromise().then(response => {
      this._notificationService.showSuccessMessage()
      return response.objResult
    })

  }
}

File: Backend/Microservices/Bursary/Eefa.Bursary.Application/Commands/ChequeSheet/Add/CreateChequeSheetCommand.cs
using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using AutoMapper;
using Eefa.Bursary.Domain.Entities.Definitions;
using Eefa.Common;
using Eefa.Common.CommandQuery;
using Eefa.Common.Data;
using MediatR;
using Microsoft.EntityFrameworkCore;

namespace Eefa.Bursary.Application.Commands.ChequeSheet
{
    public class CreateChequeSheetCommand : Common.CommandQuery.CommandBase, IRequest<ServiceResult>, IMapFrom<CreateChequeSheetCommand>, ICommand
    {
        public int Id { get; set; }
        public int? PayChequeId { get; set; }
        public int SheetSeqNumber { get; set; }
        public string SheetUniqueNumber { get; set; }
        public int SheetSeriNumber { get; set; }
        public decimal TotalCost { get; set; }
        public int? AccountReferenceId { get; set; }
        public int BankBranchId { get; set; }
        public DateTime IssueDate { get; set; }
        public DateTime ReceiptDate { get; set; }
        public int IssuerEmployeeId { get; set; }
        public string Description { get; set; }
        public string AccountNumber { get; set; }
        public string BranchName { get; set; }
        public int? OwnerChequeReferenceId { get; set; }
        public int? OwnerChequeReferenceGroupId { get; set; }
        public int? ReceiveChequeReferenceId { get; set; }
        public int? ReceiveChequeReferenceGroupId { get; set; }
        public bool? ApproveReceivedChequeSheet { get; set; }
        public int BankId { get; set; }
        public string BankCode { get; set; }
        public bool IsActive { get; set; } = true;
        public bool IsUsed { get; set; } 

        public int ChequeTypeBaseId { get; set; }

        public void Mapping(Profile profile)
        {
            profile.CreateMap<CreateChequeSheetCommand, Domain.Entities.ChequeSheets>().IgnoreAllNonExisting();
        }

    }

    public class CreateChequeSheetCommandHandler : IRequestHandler<CreateChequeSheetCommand, ServiceResult>
    {
        private readonly IMapper _autoMaper;
        private readonly IRepository<Domain.Entities.ChequeSheets> _chequeSheetRepository;
        private readonly IRepository<Banks> _bankRepository;
        private readonly IApplicationLogs _applicationLogs;
        public CreateChequeSheetCommandHandler(IMapper autoMaper, IRepository<Domain.Entities.ChequeSheets> chequeSheetRepository, IRepository<Banks> bankRepository, IApplicationLogs applicationLogs)
        {
            _autoMaper = autoMaper;
            _chequeSheetRepository = chequeSheetRepository;
            _bankRepository = bankRepository;
            _applicationLogs = applicationLogs;
        }

        public async Task<ServiceResult> Handle(CreateChequeSheetCommand request, CancellationToken cancellationToken)
        {
            await _applicationLogs.CommitLog(request);

            var bankId = await _bankRepository.GetAll().Where(x => x.Code == request.BankCode).Select(x=>x.Id).FirstAsync();
            request.BankId = bankId;
            var chequeSheet = _autoMaper.Map<Domain.Entities.ChequeSheets>(request);

            var entity = _chequeSheetRepository.Insert(chequeSheet);
            entity.ChequeDocumentState = 0;

            await _chequeSheetRepository.SaveChangesAsync(cancellationToken);

            return ServiceResult.Success();
        }

    }

}
